lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -3039,6 +3039,58 @@ bool TypeChecker::checkedCastMaySucceed(Type t1, Type t2, DeclContext *dc) {
   return (kind != CheckedCastKind::Unresolved);
 }
 
+Expr *TypeChecker::addImplicitLoadExpr(
+    Expr *expr,
+    std::function<Type(Expr *)> getType,
+    std::function<void(Expr *, Type)> setType) {
+  class LoadAdder : public ASTWalker {
+  private:
+    using GetTypeFn = std::function<Type(Expr *)>;
+    using SetTypeFn = std::function<void(Expr *, Type)>;
+
+    TypeChecker &TC;
+    GetTypeFn getType;
+    SetTypeFn setType;
+
+  public:
+    LoadAdder(TypeChecker &TC, GetTypeFn getType, SetTypeFn setType)
+      : TC(TC), getType(getType), setType(setType) {}
+
+    std::pair<bool, Expr *> walkToExprPre(Expr *E) override {
+      if (isa<ParenExpr>(E) || isa<ForceValueExpr>(E))
```

Comment:
Rather than making an `ASTWalker` and walking the tree rooted at the expression that is being loaded, why not just recurse into `ParenExpr`/`ForceValueExpr` directly, insert the load at the appropriate location, and update the subExpr on the way out?

That seems like it accomplishes the same thing without so much boilerplate.

Reply:
That's what my original version had, but then I realized that `ASTWalker` handles all of that for us e.g. recursion and `setSubExpr` so I only need to call `setType` and be done :)

Reply:
okay!